drivers/dac/ad5460/ad5460.c

```diff
@@ -0,0 +1,637 @@
+				 AD5460_SLEW_EN_MSK, 0);
+}
+
+/**
+ * @brief Enable or disable the higher thermal reset.
+ * @param desc - The device structure.
+ * @param enable - The thermal reset status.
+ * 			false: reset at 110 deg. Celsius.
+ * 			true: reset at 140 deg. Celsius.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5460_set_therm_rst(struct ad5460_desc *desc, bool enable)
+{
+	return ad5460_reg_write(desc, AD5460_THERM_RST, enable);
+}
+
+/**
+ * @brief Perform software or hardware reset and wait for device reset time.
+ * @param desc - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5460_reset(struct ad5460_desc *desc)
+{
+	int ret;
+
+	if (desc->reset_gpio) {
+		ret = no_os_gpio_direction_output(desc->reset_gpio,
+						  NO_OS_GPIO_LOW);
+		if (ret)
+			return ret;
+
+		/* Minimum RESET signal pulse duration */
+		no_os_udelay(50);
+		ret = no_os_gpio_set_value(desc->reset_gpio, NO_OS_GPIO_HIGH);
+		if (ret)
+			return ret;
+	} else {
+		ret = ad5460_reg_write(desc, AD5460_CMD_KEY, AD5460_CMD_KEY_RESET_1);
+		if (ret)
+			return ret;
+
+		ret = ad5460_reg_write(desc, AD5460_CMD_KEY, AD5460_CMD_KEY_RESET_2);
+		if (ret)
+			return ret;
+	}
+	/* Time taken for device reset (datasheet value = 1ms) */
+	no_os_mdelay(1);
+
+	return 0;
+}
+
+/**
+ * @brief Comm test function
+ * @param desc - The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int ad5460_scratch_test(struct ad5460_desc *desc)
+{
+	int ret;
+	uint16_t val;
+	uint16_t test_val = 0x1234;
+
+	ret = ad5460_reg_write(desc, AD5460_SCRATCH(0), test_val);
+	if (ret)
+		return ret;
+
+	ret = ad5460_reg_read(desc, AD5460_SCRATCH(0), &val);
+	if (ret)
+		return ret;
+
+	if (val != test_val)
+		return -EINVAL;
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the device structure.
+ * @param desc - The device structure to be initialized.
+ * @param init_param - Initialization parameter for the device descriptor.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad5460_init(struct ad5460_desc **desc,
+		struct ad5460_init_param *init_param)
+{
+	int ret;
+	struct ad5460_desc *descriptor;
+
+	if (!init_param)
+		return -EINVAL;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&descriptor->spi_desc, &init_param->spi_ip);
+	if (ret)
+		goto err;
+
+	descriptor->dev_addr = init_param->dev_addr;
+
+	no_os_crc8_populate_msb(_crc_table, AD5460_CRC_POLYNOMIAL);
+
+	ret = no_os_gpio_get_optional(&descriptor->reset_gpio,
+				      init_param->reset_gpio_param);
+	if (ret)
+		goto comm_err;
+
+	ret = ad5460_reset(descriptor);
+	if (ret)
+		goto gpio_err;
+
+	ret = ad5460_scratch_test(descriptor);
+	if (ret)
+		goto gpio_err;
+
+	*desc = descriptor;
+
+	return 0;
+
+gpio_err:
+	no_os_gpio_remove(descriptor->reset_gpio);
+comm_err:
+	no_os_spi_remove(descriptor->spi_desc);
+err:
```

Comment:
just my opinion here. these labels are not really needed... since all remove functions check for null pointer before attempting to remove. one label would suffice in my opinion, and maybe it will make the code easier to follow?